File: hexameter.core/src/main/java/org/codetome/hexameter/core/api/defaults/DefaultHexagonDataStorage.java
package org.codetome.hexameter.core.api.defaults;

import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.contract.HexagonDataStorage;
import org.codetome.hexameter.core.backport.Optional;

import java.util.LinkedHashMap;
import java.util.Map;

public class DefaultHexagonDataStorage implements HexagonDataStorage<DefaultSatelliteData> {

    private Map<CubeCoordinate, Optional<DefaultSatelliteData>> storage = new LinkedHashMap<>();

    @Override
    public final void addCoordinate(final CubeCoordinate cubeCoordinate) {
        storage.put(cubeCoordinate, Optional.<DefaultSatelliteData>empty());
    }

    @Override
    public final boolean addCoordinate(final CubeCoordinate cubeCoordinate, final DefaultSatelliteData satelliteData) {
        final Optional<DefaultSatelliteData> previous = storage.put(cubeCoordinate, Optional.of(satelliteData));
        return previous != null;
    }

    @Override
    public final Optional<DefaultSatelliteData> getSatelliteDataBy(final CubeCoordinate cubeCoordinate) {
        return storage.containsKey(cubeCoordinate) ? storage.get(cubeCoordinate) : Optional.<DefaultSatelliteData>empty();
    }

    @Override
    public final boolean containsCoordinate(final CubeCoordinate cubeCoordinate) {
        return storage.containsKey(cubeCoordinate);
    }

    @Override
    public final boolean hasDataFor(final CubeCoordinate cubeCoordinate) {
        return storage.containsKey(cubeCoordinate) && storage.get(cubeCoordinate).isPresent();
    }

    @Override
    public final Iterable<CubeCoordinate> getCoordinates() {
        return storage.keySet();
    }

    @Override
    public final boolean clearDataFor(final CubeCoordinate cubeCoordinate) {
        boolean result = false;
        if (hasDataFor(cubeCoordinate)) {
            result = true;
        }
        storage.put(cubeCoordinate, Optional.<DefaultSatelliteData>empty());
        return result;
    }
}

File: mixite.core/core/src/main/kotlin/org/hexworks/mixite/core/api/defaults/DefaultHexagonDataStorage.kt
package org.hexworks.mixite.core.api.defaults

import org.hexworks.cobalt.datatypes.Maybe
import org.hexworks.mixite.core.api.CubeCoordinate
import org.hexworks.mixite.core.api.contract.HexagonDataStorage
import org.hexworks.mixite.core.api.contract.SatelliteData

class DefaultHexagonDataStorage<T : SatelliteData> : HexagonDataStorage<T> {

    private val storage = LinkedHashMap<CubeCoordinate, Maybe<T>>()

    override val coordinates: Iterable<CubeCoordinate>
        get() = storage.keys

    override fun addCoordinate(cubeCoordinate: CubeCoordinate) {
        storage[cubeCoordinate] = Maybe.empty()
    }

    override fun addCoordinate(cubeCoordinate: CubeCoordinate, satelliteData: T): Boolean {
        val previous = storage.put(cubeCoordinate, Maybe.of(satelliteData))
        return previous != null
    }

    override fun getSatelliteDataBy(cubeCoordinate: CubeCoordinate): Maybe<T> {
        return if (storage.containsKey(cubeCoordinate)) storage[cubeCoordinate]!! else Maybe.empty()
    }

    override fun containsCoordinate(cubeCoordinate: CubeCoordinate): Boolean {
        return storage.containsKey(cubeCoordinate)
    }

    override fun hasDataFor(cubeCoordinate: CubeCoordinate): Boolean {
        return storage.containsKey(cubeCoordinate) && storage[cubeCoordinate]!!.isPresent
    }

    override fun clearDataFor(cubeCoordinate: CubeCoordinate): Boolean {
        var result = false
        if (hasDataFor(cubeCoordinate)) {
            result = true
        }
        storage[cubeCoordinate] = Maybe.empty()
        return result
    }
}
